From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Artyom Shalkhakov <artyom.shalkhakov@gmail.com>
Date: Thu, 24 Sep 2026 12:00:00 +0500
Subject: [PATCH] Decode date pickers from a xib

The xib loader (GSXib5KeyedUnarchiver) knows nothing about NSDatePicker:
a xib describes a date picker with attributes on its cell
(datePickerStyle, datePickerMode, useCurrentDate), a
<datePickerElements> element and <date> elements carrying
timeIntervalSinceReferenceDate, and none of those reach the archive keys
NSDatePickerCell reads. Every date picker loaded from a xib comes up
with no fields (elements 0) and a date two seconds into 2001.

Add decoders for NSDatePickerElements, NSDatePickerType,
NSDatePickerMode, NSDateValue (the current date when the xib says
useCurrentDate), NSMinDate and NSMaxDate.
---
diff --git a/Source/GSXib5KeyedUnarchiver.m b/Source/GSXib5KeyedUnarchiver.m
index 0a3fc50da..5da227a11 100644
--- a/Source/GSXib5KeyedUnarchiver.m
+++ b/Source/GSXib5KeyedUnarchiver.m
@@ -42,6 +42,7 @@
 #import "AppKit/NSBrowser.h"
 #import "AppKit/NSBrowserCell.h"
 #import "AppKit/NSButtonCell.h"
+#import "AppKit/NSDatePickerCell.h"
 #import "AppKit/NSCell.h"
 #import "AppKit/NSClipView.h"
 #import "AppKit/NSFormCell.h"
@@ -357,6 +358,7 @@ static NSArray      *XmlBoolDefaultYes  = nil;
                                     @"NSCellPrototype", @"NSBrFlags", @"NSNumberOfVisibleColumns",
                                     @"NSWhite", @"NSRGB", @"NSCYMK",
                                     @"NSCellFlags", @"NSCellFlags2",
+                                    @"NSDatePickerMode", @"NSDateValue",
                                     @"NSButtonFlags", @"NSButtonFlags2",
                                     @"NSUsesItemFromMenu",
                                     @"NSNormalImage", @"NSAlternateImage",
@@ -456,6 +458,12 @@ static NSArray      *XmlBoolDefaultYes  = nil;
                @"decodeTransitionStyle:", @"NSTransitionStyle",
 	       @"decodeShadowOffsetHoriz:", @"NSShadowHoriz",
 	       @"decodeShadowOffsetVert:", @"NSShadowVert",
+               @"decodeDatePickerElementsForElement:", @"NSDatePickerElements",
+               @"decodeDatePickerStyleForElement:", @"NSDatePickerType",
+               @"decodeDatePickerModeForElement:", @"NSDatePickerMode",
+               @"decodeDatePickerDateValueForElement:", @"NSDateValue",
+               @"decodeDatePickerMinDateForElement:", @"NSMinDate",
+               @"decodeDatePickerMaxDateForElement:", @"NSMaxDate",
                  nil];
           RETAIN(XmlKeyToDecoderSelectorMap);
 
@@ -2566,6 +2574,92 @@ didStartElement: (NSString*)elementName
   return object;
 }
 
+/* Date pickers.  A xib describes one with attributes on its cell
+   (datePickerStyle, datePickerMode, useCurrentDate), a <datePickerElements>
+   element naming the fields shown, and <date> elements carrying
+   timeIntervalSinceReferenceDate -- none of which match the archive keys
+   NSDatePickerCell reads, so without these every date picker from a xib
+   came up with no fields and a bogus date. */
+- (NSDate *) _dateForElement: (GSXibElement*)element key: (NSString*)key
+{
+  GSXibElement *date = [element elementForKey: key];
+  NSString *interval = [date attributeForKey: @"timeIntervalSinceReferenceDate"];
+
+  if (interval == nil)
+    return nil;
+  return [NSDate dateWithTimeIntervalSinceReferenceDate: [interval doubleValue]];
+}
+
+- (id) decodeDatePickerElementsForElement: (GSXibElement*)element
+{
+  GSXibElement *elements = [element elementForKey: @"datePickerElements"];
+  NSUInteger flags = 0;
+
+  if (elements == nil)
+    {
+      /* The Cocoa default: the date and the time, to the second. */
+      flags = NSYearMonthDayDatePickerElementFlag
+        | NSHourMinuteSecondDatePickerElementFlag;
+    }
+  else
+    {
+      if ([[elements attributeForKey: @"day"] boolValue])
+        flags |= NSYearMonthDayDatePickerElementFlag;
+      else if ([[elements attributeForKey: @"month"] boolValue]
+               || [[elements attributeForKey: @"year"] boolValue])
+        flags |= NSYearMonthDatePickerElementFlag;
+      if ([[elements attributeForKey: @"second"] boolValue])
+        flags |= NSHourMinuteSecondDatePickerElementFlag;
+      else if ([[elements attributeForKey: @"hour"] boolValue]
+               || [[elements attributeForKey: @"minute"] boolValue])
+        flags |= NSHourMinuteDatePickerElementFlag;
+      if ([[elements attributeForKey: @"timeZone"] boolValue])
+        flags |= NSTimeZoneDatePickerElementFlag;
+      if ([[elements attributeForKey: @"era"] boolValue])
+        flags |= NSEraDatePickerElementFlag;
+    }
+  return [NSNumber numberWithUnsignedInteger: flags];
+}
+
+- (id) decodeDatePickerStyleForElement: (GSXibElement*)element
+{
+  NSString *style = [element attributeForKey: @"datePickerStyle"];
+  NSUInteger value = NSTextFieldAndStepperDatePickerStyle;
+
+  if ([@"clockAndCalendar" isEqualToString: style])
+    value = NSClockAndCalendarDatePickerStyle;
+  else if ([@"textField" isEqualToString: style])
+    value = NSTextFieldDatePickerStyle;
+  return [NSNumber numberWithUnsignedInteger: value];
+}
+
+- (id) decodeDatePickerModeForElement: (GSXibElement*)element
+{
+  NSString *mode = [element attributeForKey: @"datePickerMode"];
+
+  return [NSNumber numberWithUnsignedInteger:
+    [@"range" isEqualToString: mode] ? NSRangeDateMode : NSSingleDateMode];
+}
+
+- (id) decodeDatePickerDateValueForElement: (GSXibElement*)element
+{
+  /* "Use current date" in Interface Builder: the date recorded in the
+     xib is only when it was saved. */
+  if ([[element attributeForKey: @"useCurrentDate"] boolValue])
+    return [NSDate date];
+  return [self _dateForElement: element key: @"date"];
+}
+
+- (id) decodeDatePickerMinDateForElement: (GSXibElement*)element
+{
+  return [self _dateForElement: element key: @"minDate"];
+}
+
+- (id) decodeDatePickerMaxDateForElement: (GSXibElement*)element
+{
+  return [self _dateForElement: element key: @"maxDate"];
+}
+
 - (id) decodeButtonFlags1ForElement: (GSXibElement*)element
 {
   NSNumber *value = nil;
